RFC-0011-mixed-case-identifier-support.md

```diff
@@ -0,0 +1,220 @@
+- Ensure Access Control SPI can correctly normalize identifiers.
+- Preserve a seamless user experience while making these changes.
+
+### Proposed Plan
+
+Connectors handle identifiers in three ways:
+
+1. SQL-Compliant (e.g., Oracle, DB2)
+  - Delimited identifiers (e.g., "MyTable") keep their original case.
+  - Non-delimited identifiers (e.g., mytable) are converted to uppercase.
+
+2. Case-Sensitive (e.g., PostgreSQL)
+  - Delimited identifiers keep their original case.
+  - Non-delimited identifiers may be converted to lowercase or another case.
+
+3. Case-Insensitive (e.g., Hive)
+  - Delimited and non-delimited identifiers are treated the same.
+  - Identifiers may be automatically converted to a specific case.
+
+Presto's behavior includes:
+
+- Delimited identifiers ("Identifier") and non-delimited identifiers (identifier) are converted to lowercase by default
+  unless a connector enforces a specific behavior.
+- Identifiers are normalized when:
+  - Resolving schemas, tables, columns, views.
+  - Retrieving metadata from connectors.
+  - Displaying entity names in metadata introspection commands like SHOW TABLES and DESCRIBE.
+
+Presto handles identifiers in several ways:
+
+  - Matching identifiers to locate entities such as catalogs, schemas, tables, views.
+  - Resolving column names based on table metadata provided by connectors.
+  - Passing identifiers to connectors when creating new entities.
+  - Processing and displaying entity names retrieved from connectors, including column resolution and metadata introspection commands like SHOW and DESCRIBE.
+
+## Proposed Implementation
+
+#### Core Changes
+
+* In the presto-spi, add new API to pass original identifier (Schema, table and column names)
+* Introduce new API in Metadata for preserving lower case identifier by default to preserve backward compatibility
+* Introduce new Connector specific API in ConnectorMetadata
+
+Metadata.java
+
+```java
+    String normalizeIdentifier(Session session, String catalogName, String identifier, boolean delimited);
+```
+
+MetadataManager.java
+
+```java
+    @Override
+    public String normalizeIdentifier(Session session, String catalogName, String identifier, boolean delimited)
+    {
+        Optional<CatalogMetadata> catalogMetadata = getOptionalCatalogMetadata(session, transactionManager, catalogName);
+        if (catalogMetadata.isPresent()) {
+            ConnectorId connectorId = catalogMetadata.get().getConnectorId();
+            ConnectorMetadata metadata = catalogMetadata.get().getMetadataFor(connectorId);
+            return metadata.normalizeIdentifier(session.toConnectorSession(connectorId), identifier, delimited);
+        }
+        return identifier.toLowerCase(ENGLISH);
+    }
+```
+
+ConnectorMetadata.java
+```java
+
+    /**
+     * Normalize the provided SQL identifier according to connector-specific rules
+    */
+    default String normalizeIdentifier(ConnectorSession session, String identifier, boolean delimited)
```

Comment:
Consider expanding the javadoc here to include the parameters and their descriptions (specifically I am concerned about documenting the meaning of `delimited`. I actually think we may want to rename this to "escaped" or maybe "quoted" as discussed in the conversation with Anant

Reply:
Ok, we can do that. This is going to be a boolean for `double quotes`. And if in any connector we want to handle with and without quote differently https://github.com/prestodb/rfcs/pull/36/files#diff-217aa5922002c9dabc8df724c3930f3fa969a98881495ad2648be620fdb50407R30

I chose `delimited` currently since Identifier here defines it as `delimited` https://github.com/prestodb/presto/blob/master/presto-parser/src/main/java/com/facebook/presto/sql/tree/Identifier.java#L57 But I am open for suggestions for naming.

Reply:
Delimited is widely used in other databases too e.g. [Netezza](https://www.ibm.com/docs/en/iis/11.7?topic=connector-netezza-identifiers)
But, Agree with Zac about documenting the meaning of delimited, and how it is a better alternative to escaped. Where each special character needs to be escaped and in case of delimited only the delimiter needs to be escaped.

Reply:
I think we shoun't call it escaped since this is going to be only for `double quotes` boolean. May be we can rename boolean to `quoted` if that makes more clear? Open to hear suggestions here?